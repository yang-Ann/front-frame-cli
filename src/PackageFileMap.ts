import { resolve, parse } from "node:path";

import fs from "fs-extra";
import packageFileMapConfig from "./config/packageFileMap.js";
import { getDirname } from "./utils/exports.js";

const __dirname = getDirname(import.meta.url);

// 管理依赖文件映射
export default class PackageFileMap {

  constructor(
    public projectDir: string,
    public lang: LanguageType,
    public frame: FrameType,
    public packages: Array<PackagesType>,
  ) {

  }

  private get fileMap() {
    return packageFileMapConfig.fileMap;
  }

  // 获取选中的 package (参数兼容测试用例)
  public getSelectPackages(customDir = __dirname): PackageFileMapType[] {
    const { fileMap, frame, packages, projectDir, lang } = this;

    // 从 dist 目录开始, 所有是 ../../
    const packagefromDir = resolve(customDir, `../../assets/template/${frame}/packages`);

    const result: PackageFileMapType[] = [];

    // 框架依赖 + 通用依赖
    const allPackages: ObjectType = {
      ...fileMap[frame],
      ...fileMap.commonPackageMap
    }

    Object.keys(allPackages).forEach((pn: any) => {
      if (packages.includes(pn)) {
        const value = allPackages[pn];

        // 简写字符串格式
        if (typeof value === "string") {
          result.push({
            from: build(packagefromDir, `${pn}/${value}`, true),
            to: build(projectDir, value)
          });

          // 数组格式
        } else if (Array.isArray(value)) {
          value.forEach((e: string | PackageFileMapType) => {
            // 简写字符串格式
            if (typeof e === "string") {
              result.push({
                from: build(packagefromDir, `${pn}/${e}`, true),
                to: build(projectDir, e)
              });

              // 对象格式
            } else if (e instanceof Object) {
              result.push({
                from: build(packagefromDir, `${pn}/${e.from}`, true),
                to: build(projectDir, e.to)
              });
            }
          });
        }
      }
    });

    type funcType = (lan: LanguageType) => string;
    function build(fromDir: string, path: string | funcType, isFrom = false): string {
      let result = "";

      // 是函数则调用并覆盖
      if (typeof path === "function") path = path(lang);

      result = resolve(fromDir, path);

      // 校验模板文件是否存在, 不存在则去公共目录中获取
      if (isFrom && !fs.existsSync(result)) {
        result = resolve(customDir, "../../assets/template/common/packages", path);
      }

      return result;
    }

    result
      .filter(e => {
        const { base, ext } = parse(e.to);
        // 不是 ignore 文件, 目标地址是目录
        return !base.startsWith(".") && !base.endsWith("ignore") && !ext;
      })
      .forEach(e => {
        const fromParse = parse(e.from);
        // 目标路径是目录则拼接上源文件名
        e.to = resolve(e.to, fromParse.base);
      });


    return result;
  }
}