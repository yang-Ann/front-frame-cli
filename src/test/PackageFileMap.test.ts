import { test, expect, describe } from "vitest";
import { resolve } from "node:path";

import PackageFileMap from "../PackageFileMap.js";
import {
  getDirname
} from "../utils/exports.js";

import {
  nativePackageInstance,
  reactPackageInstance,
  vuePackageInstance
} from "./const.js";

// 测试目录
const testDir = resolve(getDirname(import.meta.url));
const projectName = "C:\\project";

// 目录位置不准确
describe("getSelectPackages", () => {
  test("Vue", () => {
    const instance = new PackageFileMap(
      projectName,
      vuePackageInstance.language,
      vuePackageInstance.frame,
      vuePackageInstance.packages
    );
    expect(
      replaceSEP(instance.getSelectPackages(testDir))
    ).toMatchSnapshot("Vue 依赖映射");
  });


  test("React", () => {
    const instance = new PackageFileMap(
      projectName,
      reactPackageInstance.language,
      reactPackageInstance.frame,
      reactPackageInstance.packages
    );
    expect(
      replaceSEP(instance.getSelectPackages(testDir))
    ).toMatchSnapshot("React 依赖映射");
  });

  test("Native", () => {
    const instance = new PackageFileMap(
      projectName,
      nativePackageInstance.language,
      nativePackageInstance.frame,
      nativePackageInstance.packages
    );
    expect(
      replaceSEP(instance.getSelectPackages(testDir))
    ).toMatchSnapshot("Native 依赖映射");
  });
});


function replaceSEP(list: PackageFileMapType[]): PackageFileMapType[] {
  const RE = /\\/g;
  list.forEach(e => {
    e.from = e.from.replace(RE, "/");
    e.to = e.to.replace(RE, "/");
  });
  return list;
}