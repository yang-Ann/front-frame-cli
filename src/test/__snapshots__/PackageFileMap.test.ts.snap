// Vitest Snapshot v1

exports[`getSelectPackages > Native > Native 依赖映射 1`] = `
[
  {
    "from": "D:/project/front-frame-cli/assets/template/Native/packages/Rollup/.babelrc",
    "to": "C:/project/.babelrc",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/Native/packages/Rollup/.babelignore",
    "to": "C:/project/.babelignore",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/Native/packages/Rollup/rollup.config.ejs",
    "to": "C:/project/rollup.config.ejs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/Native/packages/Vitest/vitest.config.ejs",
    "to": "C:/project/vitest.config.ejs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/Native/packages/Vitest/test/index.test.ts",
    "to": "C:/project/src/test/index.test.ts",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Axios/request",
    "to": "C:/project/src/request",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Prettier/.prettierrc.cjs",
    "to": "C:/project/.prettierrc.cjs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Prettier/.prettierignore",
    "to": "C:/project/.prettierignore",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/Native/packages/Eslint/.eslintrc.ejs",
    "to": "C:/project/.eslintrc.ejs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/Native/packages/Eslint/.eslintignore",
    "to": "C:/project/.eslintignore",
  },
]
`;

exports[`getSelectPackages > React > React 依赖映射 1`] = `
[
  {
    "from": "D:/project/front-frame-cli/assets/template/React/packages/Reducx/store",
    "to": "C:/project/src/store",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/React/packages/Reducx/react-redux.svg",
    "to": "C:/project/src/assets/react-redux.svg",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/React/packages/Reducx/TestRedux.ejs",
    "to": "C:/project/src/components/TestRedux.ejs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/React/packages/React-router/router",
    "to": "C:/project/src/router",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/React/packages/React-router/NotFound.ejs",
    "to": "C:/project/src/components/NotFound.ejs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/React/packages/React-router/Router.ejs",
    "to": "C:/project/src/components/Router.ejs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Axios/request",
    "to": "C:/project/src/request",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Prettier/.prettierrc.cjs",
    "to": "C:/project/.prettierrc.cjs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Prettier/.prettierignore",
    "to": "C:/project/.prettierignore",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/React/packages/Eslint/.eslintrc.ejs",
    "to": "C:/project/.eslintrc.ejs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Eslint/.eslintignore",
    "to": "C:/project/.eslintignore",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Tailwind CSS/tailwind.config.cjs",
    "to": "C:/project/tailwind.config.cjs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Tailwind CSS/postcss.config.cjs",
    "to": "C:/project/postcss.config.cjs",
  },
]
`;

exports[`getSelectPackages > Vue > Vue 依赖映射 1`] = `
[
  {
    "from": "D:/project/front-frame-cli/assets/template/Vue/packages/Pinia/store",
    "to": "C:/project/src/store",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/Vue/packages/Pinia/pinia.svg",
    "to": "C:/project/src/assets/pinia.svg",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/Vue/packages/Pinia/TestPinia.ejs",
    "to": "C:/project/src/components/TestPinia.ejs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/Vue/packages/Vue-router/router",
    "to": "C:/project/src/router",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/Vue/packages/Vue-router/RouterView.ejs",
    "to": "C:/project/src/components/RouterView.ejs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/Vue/packages/Vue-router/NotFound.ejs",
    "to": "C:/project/src/components/NotFound.ejs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/Vue/packages/JSX(TSX)/TestTsx.ejs",
    "to": "C:/project/src/components/TestTsx.ejs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Axios/request",
    "to": "C:/project/src/request",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Prettier/.prettierrc.cjs",
    "to": "C:/project/.prettierrc.cjs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Prettier/.prettierignore",
    "to": "C:/project/.prettierignore",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/Vue/packages/Eslint/.eslintrc.ejs",
    "to": "C:/project/.eslintrc.ejs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Eslint/.eslintignore",
    "to": "C:/project/.eslintignore",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Tailwind CSS/tailwind.config.cjs",
    "to": "C:/project/tailwind.config.cjs",
  },
  {
    "from": "D:/project/front-frame-cli/assets/template/common/packages/Tailwind CSS/postcss.config.cjs",
    "to": "C:/project/postcss.config.cjs",
  },
]
`;
